internal/util/prometheus.go

```diff
@@ -0,0 +1,55 @@
+package util
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+type ControllerFunctionalPrometheusMetrics struct {
+	// number of post /config to proxy successfully
+	ConfigPassCounter prometheus.Counter
+	// number of post /config to proxy failed
+	ConfigFailureCounter prometheus.Counter
+	// number of ingress analysis failure
+	ParseFailureCounter prometheus.Counter
+
+	// duration of last successful confiuration sync
+	ConfigureDurationHistogram prometheus.Histogram
+}
+
+func (ctrlMetrics *ControllerFunctionalPrometheusMetrics) GeneratePrometheusCounter(name, help string) prometheus.Counter {
+	return prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: name,
+			Help: help,
+		},
+	)
+}
+
+func (ctrlMetrics *ControllerFunctionalPrometheusMetrics) GeneratePrometheusHistogram(name, help string) prometheus.Histogram {
+	return prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    name,
+			Help:    help,
+			Buckets: prometheus.LinearBuckets(20, 5, 5),
```

Comment:
For operation duration, exponential buckets (with a small factor, like `1.2`) typically work better than linear buckets. Make sure that the buckets accurately capture reasonable values that this metric will store.

Reply:
Changed per request.

Reply:
This will create only 4 buckets (+ inf) with 1, 10, 100 and 1000ms:
- I expect real-world values to lie in the 1000+ms range, so we need to scale out further in that direction
- what is the reason behind factor 10? I proposed 1.2.

What we need (I think) is:
- a possibility to capture larger durations, up to, say, 5 minutes
- more resolution than "this value lies between X and 10X"

Therefore my suggestion is to choose a factor that provides better resolution (I proposed 1.2) and adjust the number of buckets accordingly.

Reply:
agree to the 1.2 factor 👍. How about adjust to 20 buckets ? otherwise, kindly do "commit change" suggestion ?

Reply:
changed d0d14553f8bacff143d3319560b102f080d1022a